### ISSUE TITLE
Nix flake at HEAD needs new deps

### ISSUE THREAD

**Issue by alerque** [sile-typesetter/sile#1388]
The current Nix flake doesn't work because it relies on the upstream sile package to supply dependencies—amonge t hem the `buildInputs` that includes `luaEnv` which has a list of LuaRocks we need to run.

Since the last release (v0.12.5) on HEAD there are two new dependencies: `cldr` and `fluent`. We need to add them here, and eventually to the upstream Nix packaging repo as well for the next release.

@doronbehar I'm a bit lost here, the  file `pkgs/development/lua-modules/generated-packages.nix` does not look like it should be edited by hand, but I no longer see the CSV file that just listed luarocks to generate packages for, so I'm not sure how to get it refreshed with those two rocks added. Is there a tool for that? And how should we inject those two dependencties here in the flake without requiring them (yet) in the stable released package build?

**Comment by alerque** [comment]
I take that back, 3 new Lua deps: `cldr`, `fluent`, and `loadkit`.

**Comment by alerque** [comment]
Part of the reason I'm confused is that CSV file doesn't exist in my checkout. Refreshing with a pull and I even see changes to it, but still no file exists. I think my checkout might be borked ... I'll work on that and thanks for the tips.

**Comment by alerque** [comment]
Sure enough, `git fsck` turned up some file system problems.

I added the new deps and regenerated the package list, then committed just the most relevant bits. I hope that's the right track.

**Comment by alerque** [comment]
The upstream changes have landed (including the rollback to Lua 5.3 by default and new deps). I've bumped the local flake lock file twice now and still no love. What else am I missing?

**Comment by doronbehar** [comment]
Yea I'm sorry but I forgot a small detail, you need to update the `luaEnv` in nixpkgs as well. If you only want to test things locally with sile's flake, you can override the `buildInputs` attribute of sile with functions like [`lib.lists.take`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.lists.take) and remove the current `luaEnv` and add a new `luaEnv` derivation to the inputs that contains the new packages you added in https://github.com/NixOS/nixpkgs/pull/172341 . You can also update sile along with the new packages to it's `luaEnv` straight in nixpkgs, which is something you could have done as part of https://github.com/NixOS/nixpkgs/pull/172341 as well.

**Comment by alerque** [comment]
I'm still kind of stuck on this. I made a stab at appending the `buildInputs` in the flake (see #1412), specifically this bit:

https://github.com/sile-typesetter/sile/blob/32c781a9e2b88a1fa4beb81b7ac123baaf6a8b20/flake.nix#L88-L94

The result is promising in that `nix run` tries to install the Lua cldr package, but it doesn't get very far:

```
error: builder for '/nix/store/76an5fvk16pkilzp7qhzcxyda0cfxwnm-lua5.2-cldr-0.2.0-0.drv' failed with exit code 1;
       last 10 log lines:
       > source root is cldr-lua-51e4760
       > patching sources
       > configuring
       > building
       > installing
       >
       > cldr 0.2.0-0 depends on lua >= 5.1 (5.2-1 provided by VM)
       > cldr 0.2.0-0 depends on penlight (dev-1 installed)
       >
       > Error: Failed producing checksum: no MD5 checker tool available, please install md5sum, openssl or md5 in your system
       For full logs, run 'nix log /nix/store/76an5fvk16pkilzp7qhzcxyda0cfxwnm-lua5.2-cldr-0.2.0-0.drv'.
error: 1 dependencies of derivation '/nix/store/5xrrdfayvqbhmxg1xbfiz791w5c52r48-lua-5.2.4-env.drv' failed to build
error: 1 dependencies of derivation '/nix/store/3xsbxwn3nw7bwb2wgk23pg2b8r38s71b-sile-0.12.5-dirty-flake.drv' failed to build
```

This might be a problem with the Lua packages in *nixpkgs* not with the flake, but I can't make heads or tails of where to fix it.

**Comment by doronbehar** [comment]
Once the nixpkgs PR is merged, and reaches `nixos-unstable` (track it here: nixpk.gs/pr-tracker.html?pr=175749 ) You will be able to use the original `inputs.nixpkgs.url` along with the rest of the patch.

**Comment by doronbehar** [comment]
This will work for you now, without any update in Nixpkgs:

```diff
diff --git i/flake.nix w/flake.nix
index 440b35e0..fa4e7c5c 100644
--- i/flake.nix
+++ w/flake.nix
@@ -36,6 +36,32 @@
       inherit (gitignore.lib) gitignoreSource;
       # https://discourse.nixos.org/t/passing-git-commit-hash-and-tag-to-build-with-flakes/11355/2
       version_rev = if (self ? rev) then (builtins.substring 0 7 self.rev) else "dirty";
+      # Prepare a different luaEnv to be used in the overridden expression
+      luaEnv = pkgs.lua5_3.withPackages(ps: with ps; [
+        cassowary
+        cosmo
+        linenoise
+        lpeg
+        lua-zlib
+        lua_cliargs
+        luaepnf
+        luaexpat
+        luafilesystem
+        luarepl
+        luasec
+        luasocket
+        luautf8
+        penlight
+        stdlib
+        vstruct
+        cldr
+        fluent
+        loadkit
+      ] ++ pkgs.lib.optionals (pkgs.lib.versionOlder pkgs.lua.luaversion "5.2") [
+        bit32
+      ] ++ pkgs.lib.optionals (pkgs.lib.versionOlder pkgs.lua.luaversion "5.3") [
+        compat53
+      ]);
       # Use the expression from Nixpkgs instead of rewriting it here.
       sile = pkgs.sile.overrideAttrs(oldAttr: rec {
         version = "${(pkgs.lib.importJSON ./package.json).version}-${version_rev}-flake";
@@ -85,6 +111,16 @@
         nativeBuildInputs = oldAttr.nativeBuildInputs ++ [
           pkgs.autoreconfHook
         ];
+        buildInputs = [
+          # NOTE: This adds a different `lua` interpreter to the `buildInputs`,
+          # in addition to the old `luaEnv` defined in nixpkgs' lua expression
+          luaEnv
+        ] ++ oldAttr.buildInputs;
+        # This is written in Nixpkgs' expression as well, but we need to write
+        # this here as well so that the overridden luaEnv will be used instead
+        passthru = {
+          inherit luaEnv;
+        };
         # TODO: This switch between the hooks can be moved to Nixpkgs'
         postPatch = oldAttr.preConfigure;
         preConfigure = "";
```

**Comment by doronbehar** [comment]
The following change, will be more suitable once https://github.com/NixOS/nixpkgs/pull/175749 lands in `nixos-unstable`, and you'd `nix flake update`.

```diff
diff --git i/flake.nix w/flake.nix
index 440b35e0..c1b5a89f 100644
--- i/flake.nix
+++ w/flake.nix
@@ -36,6 +36,33 @@
       inherit (gitignore.lib) gitignoreSource;
       # https://discourse.nixos.org/t/passing-git-commit-hash-and-tag-to-build-with-flakes/11355/2
       version_rev = if (self ? rev) then (builtins.substring 0 7 self.rev) else "dirty";
+      # Prepare a different luaEnv to be used in the overridden expression,
+      # this is also the place to choose a different lua interpreter, such as
+      # lua5_3 or luajit
+      luaEnv = pkgs.lua5_3.withPackages(ps: with ps; [
+        cassowary
+        cosmo
+        linenoise
+        lpeg
+        lua-zlib
+        lua_cliargs
+        luaepnf
+        luaexpat
+        luafilesystem
+        luarepl
+        luasec
+        luasocket
+        luautf8
+        penlight
+        stdlib
+        vstruct
+        cldr
+        fluent
+        loadkit
+        # If we want to test things with lua5.2 or an even older lua, we uncomment these
+        #bit32
+        #compat53
+      ]);
       # Use the expression from Nixpkgs instead of rewriting it here.
       sile = pkgs.sile.overrideAttrs(oldAttr: rec {
         version = "${(pkgs.lib.importJSON ./package.json).version}-${version_rev}-flake";
@@ -85,9 +112,23 @@
         nativeBuildInputs = oldAttr.nativeBuildInputs ++ [
           pkgs.autoreconfHook
         ];
-        # TODO: This switch between the hooks can be moved to Nixpkgs'
-        postPatch = oldAttr.preConfigure;
-        preConfigure = "";
+        buildInputs = [
+          # This adds a different `lua` interpreter to the `buildInputs`.
+          luaEnv
+        ] ++ (
+          # We remove the first buildInput from nixpkgs which is the luaEnv
+          # used there. It's not mandatory to do so, because anyway the first
+          # `lua` interpreter that appears in the final `buildInputs` is the
+          # the `lua` that's used in the `$PATH`, and eventually in the built
+          # `sile`, but we'd like to keep the `buildInputs` clean if possible
+          # never the less.
+          pkgs.lib.lists.drop 1 oldAttr.buildInputs
+        );
+        # This is written in Nixpkgs' expression as well, but we need to write
+        # this here so that the overridden luaEnv will be used instead.
+        passthru = {
+          inherit luaEnv;
+        };
         meta = oldAttr.meta // {
           changelog = "https://github.com/sile-typesetter/sile/raw/master/CHANGELOG.md";
         };
diff --git i/flake.nix w/flake.nix
index 440b35e0..c1b5a89f 100644
--- i/flake.nix
+++ w/flake.nix
@@ -36,6 +36,33 @@
       inherit (gitignore.lib) gitignoreSource;
       # https://discourse.nixos.org/t/passing-git-commit-hash-and-tag-to-build-with-flakes/11355/2
       version_rev = if (self ? rev) then (builtins.substring 0 7 self.rev) else "dirty";
+      # Prepare a different luaEnv to be used in the overridden expression,
+      # this is also the place to choose a different lua interpreter, such as
+      # lua5_3 or luajit
+      luaEnv = pkgs.lua5_3.withPackages(ps: with ps; [
+        cassowary
+        cosmo
+        linenoise
+        lpeg
+        lua-zlib
+        lua_cliargs
+        luaepnf
+        luaexpat
+        luafilesystem
+        luarepl
+        luasec
+        luasocket
+        luautf8
+        penlight
+        stdlib
+        vstruct
+        cldr
+        fluent
+        loadkit
+        # If we want to test things with lua5.2 or an even older lua, we uncomment these
+        #bit32
+        #compat53
+      ]);
       # Use the expression from Nixpkgs instead of rewriting it here.
       sile = pkgs.sile.overrideAttrs(oldAttr: rec {
         version = "${(pkgs.lib.importJSON ./package.json).version}-${version_rev}-flake";
@@ -85,9 +112,23 @@
         nativeBuildInputs = oldAttr.nativeBuildInputs ++ [
           pkgs.autoreconfHook
         ];
-        # TODO: This switch between the hooks can be moved to Nixpkgs'
-        postPatch = oldAttr.preConfigure;
-        preConfigure = "";
+        buildInputs = [
+          # This adds a different `lua` interpreter to the `buildInputs`.
+          luaEnv
+        ] ++ (
+          # We remove the first buildInput from nixpkgs which is the luaEnv
+          # used there. It's not mandatory to do so, because anyway the first
+          # `lua` interpreter that appears in the final `buildInputs` is the
+          # the `lua` that's used in the `$PATH`, and eventually in the built
+          # `sile`, but we'd like to keep the `buildInputs` clean if possible
+          # never the less.
+          pkgs.lib.lists.drop 1 oldAttr.buildInputs
+        );
+        # This is written in Nixpkgs' expression as well, but we need to write
+        # this here so that the overridden luaEnv will be used instead.
+        passthru = {
+          inherit luaEnv;
+        };
         meta = oldAttr.meta // {
           changelog = "https://github.com/sile-typesetter/sile/raw/master/CHANGELOG.md";
         };
```

**Comment by alerque** [comment]
@doronbehar Thanks so much for the help with this. Your help has been invaluable.

I started playing with this yesterday with your first patch, and now that the PR has landed all the way in `nixos-unstable` with the second patch today (as seen in #1412). Unfortunately in both cases I got stuck on a new issue.

I think we are past the Lua stuff: the autoconf configure script finds the new Lua deps as well as everything else it needs, and SILE builds and at least the core Lua script and the `require()` statements for our internal stuff and Lua Rocks seems to run fine.

What it dies on is loading our `justenoughicu` module. That module itself seems to be built and installed in the Nix store just fine and Lua finds it. But that module is dynamically linked against the `icu` package, specifically `libicuio.so.71`. Our configure script found ICU at build time so it seems like the `buildInputs` are correct. I also see that module in some package in the store. But for some reason Lua can't find it at SILE runtime.

Any ideas what might still be missing from this equation?

**Comment by alerque** [comment]
Huh, okay I have lots more info having tested on NixOS and some other machines. It looks like the problem is entirely localized to running the flake from a CWD that also has a compiled SILE in it. If I get out of there and use a remote Git branch instead it seems to work fine.

**Comment by alerque** [comment]
Gah! So close and yet....

It works for me locally:

```console
$ nix run github:sile-typesetter/sile -- --version
SILE v0.12.5-72b25f3-flake (Lua 5.3)
```

But [CI bombs out](https://github.com/alerque/sile-typesetter.github.io/runs/6802519093?check_suite_focus=true#step:5:783) with a cryptic:

> error: attribute 'defaultApp.x86_64-linux' should have type 'derivation'

Same version of `nix` (2.9.1) in both places, same channels (I think) and same config options (seemingly irrelevant experimental flags leftover from previous versions, but tried with and without).
